### DISCUSSION TITLE
AttributeError: 'Depends' object has no attribute 'execute' when using async sqlalchemy with fastapi

### DISCUSSION THREAD

**Discussion by Kokoserver** [fastapi/fastapi#9344] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# I created async_maker and session for fastapi dependency:
engine = create_async_engine(config.get_database_url(), echo=False)
Async_session: AsyncSession = async_sessionmaker(bind=engine, expire_on_commit=False)
and async session:
 async def get_async_session():
   async with async_maker() as session:
    yield session

# when i use this with fastapi
async def example(session: AsyncSession = Depends(get_async_session)):
 stm = select(model.Permission)
 result = await session.execute(stm)
 return result.scalars().all()


async def example():
async with async_maker() as session:
 stm = select(model.Permission)
 result = await session.execute(stm)
 return result.scalars().all()
```

### Description

I created async_maker and session for fastapi dependency:
engine = create_async_engine(config.get_database_url(), echo=False)
Async_session: AsyncSession = async_sessionmaker(bind=engine, expire_on_commit=False)
and async session:
async def get_async_session():
 async with async_maker() as session:
   yield session
when i use this with fastapi
async def example(session: AsyncSession = Depends(get_async_session)):
 stm = select(model.Permission)
 result = await session.execute(stm)
 return result.scalars().all()

i get the following error runing the above function from API endpoint call:
result = await session.execute(stm)
^^^^^^^^^^^^^^^
'Depends' object has no attribute 'execute'
but if i use the session directly like this:

async def example():
async with async_maker() as session:
 stm = select(model.Permission)
 result = await session.execute(stm)
 return result.scalars().all()
it will work fine, but during testing using pytest i will get this error:
self = <_ProactorSocketTransport fd=684 read=<_OverlappedFuture cancelled>>, f = None, data = b'Q\x00\x00\x00*BEGIN ISOLATION LEVEL READ COMMITTED;\x00'
def _loop_writing(self, f=None, data=None):
    try:
        if f is not None and self._write_fut is None and self._closing:
            # XXX most likely self._force_close() has been called, and
            # it has set self._write_fut to None.
            return
        assert f is self._write_fut
        self._write_fut = None
        self._pending_write = 0
        if f:
            f.result()
        if data is None:
            data = self._buffer
            self._buffer = None
        if not data:
            if self._closing:
                self._loop.call_soon(self._call_connection_lost, None)
            if self._eof_written:
                self._sock.shutdown(socket.SHUT_WR)
            # Now that we've reduced the buffer size, tell the
            # protocol to resume writing if it was paused.  Note that
            # we do this last since the callback is called immediately
            # and it may add more data to the buffer (even causing the
            # protocol to be paused again).
            self._maybe_resume_protocol()
        else:
          self._write_fut = self._loop._proactor.send(self._sock, data)
  result = await db.execute(stm)
                  ^^^^^^^^^^
AttributeError: 'Depends' object has no attribute 'execute'
please help

### Operating System

Windows

### FastAPI Version

3.10

### Python Version

{extras = ["all"], version = "^0.95.0"}

### Additional Context

sqlalchemy = "^2.0.7"

**Comment by DChirillov** [comment]
Have you fixed the issue?

**Reply by kamikaze** [reply]
Same here with:

```
fastapi==0.100.1
fastapi-users[sqlalchemy]==11.0.0
pydantic==1.10.12
SQLAlchemy==1.4.49
starlette==0.27.0
```

**Reply by Kokoserver** [reply]
I had to use another solution

**Reply by kamikaze** [reply]
which one?

**Comment by ricardomrcruz** [comment]
Hello I had the same issue in a auth handler function while executing a query. Depends from fastapi only works on endpoints. 

Well the cleanest way I found to solve this is to avoid the injection completly and use the session directly inside the function. 

Taking your example:

```
engine = create_async_engine(config.get_database_url(), echo=False)

SessionLocal = sessionmaker(bind=engine, expire_on_commit=False)

async def example():
async with SessionLocal() as session: 

 stm = select(model.Permission)
 result = await session.execute(stm)
 return result.scalars().all()
```

This helped me debug my code. It might be able to help you with pytest.

**Comment by AliHezarpisheh** [comment, marked as answer]
This is happening because your session parameter type in the example function is a `Depends` object, not an AsyncSession or Session object.
Outside of FastApi routers, `Depends` can not inject dependencies into the functions.

Solution: Use a path operation decorator to make the example function into a FastAPI path operation function. With this approach, the dependency injection is handled by the framework and you will get an AsyncSession type for the session parameter in the example function.

**Reply by Maksim-07** [reply]
And how to do it?
